ADR/0028-component-dependencies.md

```diff
@@ -0,0 +1,179 @@
+* [integration-service] will always skip testing for all components except for the bundle image Component, because the bundle image Component declares that it depends on all other Components. (All other Components are **dependency** Components. The bundle is the only **dependant** Component.)
+* [build-service] will propagate digest references as a PR to the bundle image, in the same PR Group
+  as the originating PR. The PR will be marked as "Draft" to indicate they should not be merged.
+* When the operand image PR is merged, [build-service] will update the PRs it originally filed to
+  take them out of "Draft" to indicate that they are safe to merge now as long as there are no other
+  unmerged triggering PRs in the same group.
+
+Think about branches:
+
+* User team updates their operand images with branch `feature-1234` (or whatever).
+* [build-service] uses that same branch in its PR to the bundle repo.
+* [build-service] uses that branch name as a prefix in the branch names that it chooses (`feature-1234/<suffix>`) to send to the bundle repo.
+* If user submits three operand image PRs and uses the same branch name for all (`feature-1234`) – then the digests all pile up on the same bundle branch (`feature-1234/<suffix>`), the same bundle PR.
+* If the user submits three operand image PRs and uses different branch names for all – then the digests are split among three different bundle branches, three different bundle PRs.
+
+---
+
+### OLM Operators, with components in the same repo
+
+Another OLM Operator Scenario: an operator git repo contains both the controller code and the bundle metadata.
+
+* The user has two Components, that both point to the same repo.
+* The bundle Component declares that it `depends on` the controller image Component, by way of the new field on the Component CR.
+* [integration-service] will always skip testing for the controller Component because it is known to be an **dependency** component.
+* [build-service] will propagate the digest reference as a PR to the bundle image, which happens to be the same git repository as the controller Component.
+
+Think about branches:
+
+* The user submitted their controller image update on the `new-feature` branch.
+* [build-service] pushes its commits with the new image digest references to a new feature branch that uses the triggering feature branch name as a prefix: `new-feature/<suffix>`.
+  * This works like it does in most other cases. Let the user merge the original PR, and only after that will [integration-service] update the checks for `new-feature` to say “okay to merge now!”
+  * When the user merges the original PR, [build-service] will update the PR it filed on the `new-feature/<suffix>` branch to take it out of "Draft", indicating that it is safe to merge now. It may potentially rebase the PR to trigger a new build or use `/retest`.
```

Comment:
Do the pull requests on the dependant components actually make sense once the dependency PRs are merged?

e.g. if I have feature A and feature B in the dependency, you will get an A and B pr to test that these features work with the dependant.

If I then merge A and B why do I still have two PRs in the dependant repo? Don't I want just one PR to update to the most recent version of the dependency? As a user if the dependency has merged multiple pull requests I will end up with multiple mergable pull requests, all modifying the same file, with no obvious way to know which is the correct one.

Reply:
What I mean is that conceptually when a downstream PR is open then having the draft open for testing kinda makes sense (although personally I think there be resistance too it, as it fills the PR queue with unmeragable PRs that are not directly related to the downstream project).

Once a PR is merged however there is effectively no PR any more, just the current state of the main branch. At most there should only ever be a single PR open that is actually mergable, which is to update to the most recent version of main (like we do in infra deployments).

Reply:
@dirgim, how might this behavior affect the cases where integration tests are run from the PUSH events?

There is a tradeoff in complexity when it comes to managing PRs. @stuartwdouglas isn't available to comment on this clarification, but I see the two situations as follows:

*NOTE: In both of the cases below, after A1 and B1 merge, C1 and C2 will need to be updated as a _new_ version of Components A and B will be build from the push event.*

### Individual mergable dependent PRs

* Component A depends on Component C; PR A1 results in the creation of PR C1
* Component B depends on Component C; PR B1 results in the creation of PR C2
* A1 merges, C1 is marked to be ready to merge
* B1 merges, C2 is marked ready to be merged
* C1 merges, C2 might need a rebase (degenerate case is B == A so there would be a merge conflict)
* C2 merges

### Single mergable dependent PR

* Component A depends on Component C; PR A1 results in the creation of PR C1
* Component B depends on Component C; PR B1 results in the creation of PR C2
* A1 merges, C1 is marked ready to merge
* B1 merges, C2 is closed and the change is applied to C1
* C1 merges

An alternative to this approach _could_ be for C1 to close as well if the patter was to _always_ open a new PR to merge. I suspect that this would be more confusing to the user though.